docs/rfc/rfc-018-bls-agg-exploration.md

```diff
@@ -0,0 +1,400 @@
+General [packets][ibc-packet] only contain enough information to correctly
+route the data to the counterparty chain.
+
+IBC does persist commit signatures to the chain in these `MsgUpdateClient`
+message when updating state. This message would no longer need the full set
+of unique signatures and would instead only need one signature for all of the
+data in the header.
+
+Adding BLS signatures would create a new signature type that must be
+understood by the IBC module and by the relayers. For some operations, such
+as state updates, the set of data written into the chain and received by the
+IBC module would be slightly smaller.
+
+## Discussion
+
+### What are the proposed benefits to aggregated signatures?
+
+#### Reduce Block Size
+
+At the moment, the commit size contains a 512 bit signature for each validator
+that voted for the block. For the cosmos hub, which has 175 validators in the
+validator set, this amounts to about 11kb per block. That is around 113 GB over
+the lifetime of the chain's 10118561 blocks. (Note, the hub has increased the
+number of validators in the set over time so the signature size over the history
+of the chain is likely less than that value). 
+
+Signature aggregation would only produce two signatures for the entire block.
+One for the yays and one for the nays. Each BLS aggregated signature is 48
+bytes, per the [IETF standard of BLS signatures][bls-ietf-ecdsa-compare].
+Over the lifetime of the same cosmos hub chain, that would amount to about 1
+GB, a savings of 112 GB. While that is a large factor of reduction it's worth
+bearing in mind that, at [GCP's cost][gcp-storage-pricing] of $.026 USD per GB,
+that is a total savings of around $2.50 per month.
+
+
+
+
+#### Reduce Signature Verification Creation and Verification Time
+
+From the [IETF draft standard on BLS Signatures][bls-ietf], BLS signatures can be
+created in 370 microseconds and verified in 2700 microseconds. Our current
+[Ed25519 implementation][voi-ed25519-perf] lists a 27.5 microsecond signature creation time
+and 5.19 milliseconds to perform batch verification on 128 signatures, which is
+slightly fewer than the 175 in the hub. blst, a popular implementation of BLS
+signature aggregation was benchmarked to perform verification on 100 signatures in
+1.5 milliseconds [when run locally][blst-verify-bench] on an 8 thread machine and
+pre-aggregated public keys.
+
+#### Reduce Light-Client Verification Time
+
+The light client aims to be a faster and lighter-weight way to verify that a 
+block was voted on by a Tendermint network. The light client fetches
+Tendermint block headers and commit signatures, performing public key
+verification to ensure that the associated validator set signed the block.
+Reducing the size of the commit signature would allow the light client to fetch
+block data more quickly.
+
+Additionally, the faster signature verification times of BLS signatures mean
```

Comment:
One could also imagine having an optional switch to have nodes compute aggregated signatures separately from consensus, for use by light clients—as far as I can tell there's no reason this has to be done ahead of time, as long as we have the individual inputs in storage somewhere. (Maybe that's wrong)

Reply:
Yes, I see no issue with this. Aggregation takes separate signatures and combines them, see [the ietf API description](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.8), so there should be no qualm with aggregating them at different points or storing them separately.